packages/tldraw/src/lib/ui/components/A11y.tsx

```diff
@@ -41,8 +52,123 @@ export function SkipToMainContent() {
 
 /** @public @react */
 export const DefaultA11yAnnouncer = memo(function TldrawUiA11yAnnouncer() {
-	const currentMsg = useA11y()
-	const msg = useValue('a11y-msg', () => currentMsg.currentMsg.get(), [])
+	const a11y = useA11y()
+	const msg = useValue('a11y-msg', () => a11y.currentMsg.get(), [])
+	useA11yDebug(msg.msg)
+
+	useSelectedShapesAnnouncer()
+
+	return (
+		msg.msg && (
+			<div
+				aria-live={msg.priority || 'assertive'}
+				role="status"
+				aria-hidden="false"
+				style={{
+					position: 'absolute',
+					top: '-10000px',
+					left: '-10000px',
+				}}
+			>
+				{msg.msg}
+			</div>
+		)
+	)
+})
+
+/**
+ * Core function to generate accessibility announcements for selected shapes
+ * @public
+ */
+export function generateShapeAnnouncementMessage(args: {
+	editor: Editor
+	selectedShapeIds: TLShapeId[]
+	msg(id: string, values?: Record<string, any>): string
+}) {
+	const { editor, selectedShapeIds, msg } = args
+	let a11yLive = ''
+	const numShapes = selectedShapeIds.length
+
+	if (numShapes > 1) {
+		a11yLive = msg('a11y.multiple-shapes').replace('{num}', numShapes.toString())
+	} else if (numShapes === 1) {
+		const shapeId = selectedShapeIds[0]
+		const shape = editor.getShape(shapeId)
+		if (!shape) return ''
+
+		const shapeUtil = editor.getShapeUtil(shape.type)
+
+		const isMedia = ['image', 'video'].includes(shape.type)
+		// Yeah, yeah this is a bit of a hack, we should get better translations.
+		let shapeType = ''
+		if (shape.type === 'geo') {
+			shapeType = msg(`geo-style.${(shape as TLGeoShape).props.geo}`)
+		} else if (isMedia) {
+			shapeType = msg(`a11y.shape-${shape.type}`)
+		} else {
+			shapeType = msg(`tool.${shape.type}`)
+		}
+
+		// Get shape index in reading order
+		const readingOrderShapes = editor.getCurrentPageShapesInReadingOrder()
+		const currentShapeIndex = (readingOrderShapes.findIndex((s) => s.id === shapeId) + 1).toString()
+		const totalShapes = readingOrderShapes.length.toString()
+		const shapeIndex = msg('a11y.shape-index')
+			.replace('{num}', currentShapeIndex)
+			.replace('{total}', totalShapes)
+
+		// Get describing text (alt text or shape text)
+		const describingText =
+			shapeUtil.getAriaDescriptor(shape) || shapeUtil.getText(shape) || msg('a11y.shape-empty')
+
+		// Build the full announcement
+		a11yLive = `${describingText}, ${shapeType}. ${shapeIndex}`
```

Comment:
instead of "Empty, star. 2 of 4" can/should we just do "Star, 2 of 4"? I think we should put the most useful/actionable info first